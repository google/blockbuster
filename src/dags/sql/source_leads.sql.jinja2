SELECT
  ga_sessions.*,
  CAST(lead_history.most_recent_lead AS INT64) AS most_recent_lead
FROM `{{ ga_source }}` AS ga_sessions
INNER JOIN (
  WITH leads AS (
    SELECT
      fullVisitorId,
      (visitStartTime + hits.time/1000) as lead_time
    FROM `{{ ga_source }}` sessions,
      UNNEST(hits) as hits
    WHERE {{ lead_filter_criteria|join(' AND ') }}
  )
  SELECT
    sessions.fullVisitorId,
    sessions.visitId,
    MAX(leads.lead_time) as most_recent_lead
  FROM `{{ ga_source }}` sessions
  LEFT OUTER JOIN leads
    ON sessions.fullVisitorId = leads.fullVisitorId
      AND sessions.visitStartTime > leads.lead_time
  WHERE _TABLE_SUFFIX >= '{{ window_start_date }}'
    AND sessions.fullVisitorId IN (SELECT fullVisitorId FROM leads)
  GROUP BY 1, 2
) AS lead_history
USING (fullVisitorId, visitId)
WHERE _TABLE_SUFFIX >= '{{ window_start_date }}'
  AND lead_history.most_recent_lead IS NOT NULL
