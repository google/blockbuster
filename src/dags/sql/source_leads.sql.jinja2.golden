SELECT
  ga_sessions.*,
  CAST(lead_history.most_recent_lead AS INT64) AS most_recent_lead
FROM `ga_project.dataset.table` AS ga_sessions
INNER JOIN (
  WITH leads AS (
    SELECT
      fullVisitorId,
      (visitStartTime + hits.time/1000) as lead_time
    FROM `ga_project.dataset.table` sessions,
      UNNEST(hits) as hits
    WHERE _TABLE_SUFFIX >= '20200101' AND hits.eventInfo.eventCategory like '%LEAD%'
  )
  SELECT
    sessions.fullVisitorId,
    sessions.visitId,
    MAX(leads.lead_time) as most_recent_lead
  FROM `ga_project.dataset.table` sessions
  LEFT OUTER JOIN leads
    ON sessions.fullVisitorId = leads.fullVisitorId
      AND sessions.visitStartTime > leads.lead_time
  WHERE _TABLE_SUFFIX >= '20191201'
    AND sessions.fullVisitorId IN (SELECT fullVisitorId FROM leads)
  GROUP BY 1, 2
) AS lead_history
USING (fullVisitorId, visitId)
WHERE _TABLE_SUFFIX >= '20191201'
  AND lead_history.most_recent_lead IS NOT NULL
